File: Chat Server/Chat Server/CommandLineArguments.c
//
//  CommandLineArguments.c
//
//  Created by Matt Zanchelli on 3/26/14.
//  Copyright (c) 2014 Matt Zanchelli. All rights reserved.
//

#include "CommandLineArguments.h"

#include "Options.h"
#include "VectorHelper.h"

int main(int argc, const char * argv[])
{
	std::string name;
	std::vector<std::string> arguments;
	std::map<std::string, bool> options;
	
	// The executable's name is the first argument.
	name = argv[0];
	
	if ( argc >= 2 ) {
		// Get the remaining arguments in a vector.
		arguments = vectorOfStringsFromArrayOfCStrings(argc-1, &argv[1]);
		// Remove the arguments that are options, and hold on to the options.
		options = getOptionsFromCommandLineArguments(&arguments);
	}
	
	// Run the program with the specified options.
	return run(name, options, arguments);
}

void printUsage()
{
	std::cout << "Usage: chat_server <port> [<port> ... <port>]" << std::endl;
}

File: Chat Server/Chat Server/CommandLineArguments.m
//
//  CommandLineArguments.m
//  Chat Server
//
//  Created by Matt Zanchelli on 3/27/14.
//  Copyright (c) 2014 Matt Zanchelli. All rights reserved.
//

#import "CommandLineArguments.h"

#import "ArrayHelper.h"
#import "Options.h"

int main(int argc, const char * argv[])
{
#if !__has_feature(objc_arc)
	// Autorelease Pool.
	NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
#endif
	
	// The objects to pass along to run().
	NSString *name;
	NSDictionary *options = [[NSDictionary alloc] init];
	NSArray *arguments = [[NSArray alloc] init];
	
	// The executable's name is the first argument.
	name = [NSString stringWithFormat:@"%s", argv[0]];
	
	if ( argc >= 2 ) {
		// Get the remaining arguments in an array.
		NSMutableArray *allArguments = [NSMutableArray arrayWithArray:arrayFromArrayOfCStrings(argc-1, &argv[1])];
		// Remove the arguments that are options, and hold on to the options.
		options = getOptionsFromCommandLineArgumentsFromArray(allArguments);
		// Immutable copy to return.
		arguments = allArguments;
	}
	
	// Run the program with the specified options.
	return run(name, options, arguments);
	
#if !__has_feature(objc_arc)
	[pool drain];
#endif
}

void printUsage()
{
	printf("Usage: chat_server <port> [<port> ... <port>]\n");
}

